internal/scorecard/alpha/scorecard.go

```diff
@@ -30,68 +30,74 @@ import (
 )
 
 type Scorecard struct {
-	Config          Config
-	Selector        labels.Selector
-	BundlePath      string
-	WaitTime        time.Duration
-	Kubeconfig      string
-	Namespace       string
-	bundleConfigMap *v1.ConfigMap
-	ServiceAccount  string
-	Client          kubernetes.Interface
-	SkipCleanup     bool
+	Config              Config
+	Selector            labels.Selector
+	BundlePath          string
+	WaitTime            time.Duration
+	Namespace           string
+	ServiceAccount      string
+	bundleConfigMapName string
+	Client              kubernetes.Interface
+	SkipCleanup         bool
 }
 
 // RunTests executes the scorecard tests as configured
 func (o Scorecard) RunTests() (testOutput v1alpha2.ScorecardOutput, err error) {
-	tests := selectTests(o.Selector, o.Config.Tests)
+	tests := o.selectTests()
 	if len(tests) == 0 {
-		fmt.Println("no tests selected")
+		testOutput.Results = make([]v1alpha2.ScorecardTestResult, 0)
 		return testOutput, err
 	}
 
-	bundleData, err := getBundleData(o.BundlePath)
+	ctx, cancel := context.WithTimeout(context.Background(), o.WaitTime)
+	defer cancel()
+
+	bundleData, err := o.getBundleData()
 	if err != nil {
 		return testOutput, fmt.Errorf("error getting bundle data %w", err)
 	}
 
-	// create a ConfigMap holding the bundle contents
-	o.bundleConfigMap, err = createConfigMap(o, bundleData)
+	err = o.CreateConfigMap(ctx, bundleData)
 	if err != nil {
 		return testOutput, fmt.Errorf("error creating ConfigMap %w", err)
 	}
 
-	for i, test := range tests {
-		var err error
-		tests[i].TestPod, err = o.runTest(test)
+	testOutput.Results = make([]v1alpha2.ScorecardTestResult, len(tests))
+
+	for idx, test := range tests {
+		result, err := o.runTest(ctx, test)
 		if err != nil {
-			return testOutput, fmt.Errorf("test %s failed %w", test.Name, err)
+			result = convertErrorToResult(test.Name, test.Description, err)
 		}
+		testOutput.Results[idx] = result
 	}
 
 	if !o.SkipCleanup {
-		defer deletePods(o.Client, tests)
-		defer deleteConfigMap(o.Client, o.bundleConfigMap)
-	}
-
-	err = o.waitForTestsToComplete(tests)
-	if err != nil {
-		return testOutput, err
+		defer func() {
+			err := o.deletePods(ctx)
+			if err != nil {
+				testOutput.Results = append(testOutput.Results,
+					convertErrorToResult("", "", err))
+			}
+			err = o.deleteConfigMap(ctx)
+			if err != nil {
+				testOutput.Results = append(testOutput.Results,
+					convertErrorToResult("", "", err))
+			}
+		}()
```

Comment:
Using `defer` here doesn't help much since we're already at the end of the function.

Also, these errors should probably be returned instead of packaged as test results since they occurred as part of setup/teardown.

In this case, since we DO have valid test results, we can return `testOutput, err`